File: mvp/src/main/java/me/giacoppo/mvp/executor/UIThread.java
package me.giacoppo.mvp.executor;

import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;

public class UIThread implements PostExecutionThread {
    private UIThread() {}

    @Override
    public Scheduler getScheduler() {
        return AndroidSchedulers.mainThread();
    }

    public static class SingletonHolder {
        private static final PostExecutionThread instance = new UIThread();

        private SingletonHolder(){}

        public static PostExecutionThread getInstance() {
            return instance;
        }
    }
}

File: mvp/src/main/java/me/giacoppo/mvp/executor/UIThread.kt
package me.giacoppo.mvp.executor

import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers

class UIThread private constructor() : PostExecutionThread {

    override fun getScheduler(): Scheduler {
        return AndroidSchedulers.mainThread()
    }

    private object Holder {
        val instance = UIThread()
    }

    companion object {
        val instance: PostExecutionThread by lazy { Holder.instance }
    }
}
